master/buildbot/db/buildsets.py

```diff
@@ -102,23 +107,6 @@ def thd(conn):
                 if row and (row.startbrid is not None):
                     startbrid = row.startbrid
 
-            # Check if anything in brDictsToMerge finished while we were being created.
-            # We do this to avoid database mrge conditions where we could be merging into a breq
-            # that has finished during this merge process (which means that our new breq would
-            # never be finished by master).
-            # Not that doing this second check here only works because we rely on the build finish
-            # code running in the same thread as we are (never in parallel, and certainly would not
-            # work in a multi-master environment).
-            mergebrids = [mergeBrDict['brid'] for mergeBrDict in brDictsToMerge.values()]
-            q = sa.select([br_tbl.c.buildername]) \
```

Comment:
Have you considered using here `SELECT ... LOCK IN SHARE MODE` (by using parameter `for_update='read'`). This will block all updates on rows selected by this query till the transaction end. (https://dev.mysql.com/doc/refman/5.6/en/innodb-locking-reads.html)

Reply:
We have two problem here:

* Ensuring that a build does not finish while we are merging into it (could be done with a row lock as you suggested)
* Ensuring that once a build is finishing, nothing else gets merged into it.

The second problem is a bit harder to do with db locks, because there are no rows to lock (they are being added in this process), so we would have to lock the entire buildrequests table.

Instead of doing that, we opted to lock only those two processes (finishing and adding new merged builds).

Does that make sense? Do you see another way to do this with db locks?

Reply:
Thanks for clarification, it makes sense 👍